Add new rule `no-invalid-dependent-keys`

TheMBTH opened the pull request:
This rule aims to avoid mismatched open and closed braces.

As discussed in the comments of this issue #105, I added a rule to check that curly braces are balanced within a computed properties' parameters

Fixes #708.

bmish commented on lib/rules/balanced-braces.js:
```diff
@@ -0,0 +1,61 @@
+    docs: {
+      description: 'require that braces to be balanced.',
+      category: 'Possible Errors',
+      recommended: true,
```
New rules should not be added as recommended, but we will enable this rule as recommended in the next major release :)

bmish commented on lib/rules/balanced-braces.js:
```diff
@@ -0,0 +1,61 @@
+      url:
+        'https://github.com/ember-cli/eslint-plugin-ember/tree/master/docs/rules/balanced-braces.md',
+    },
+    fixable: 'code',
```
The rule is not autofixable so this property should be removed.

bmish commented on lib/rules/balanced-braces.js:
```diff
@@ -0,0 +1,61 @@
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'require that braces to be balanced.',
```
Description should mention that this is referring to computed property dependent keys. Otherwise it's unclear what braces this is referring to.

bmish commented on lib/index.js:
```diff
@@ -5,6 +5,7 @@ module.exports = {
     'alias-model-in-controller': require('./rules/alias-model-in-controller'),
     'avoid-leaking-state-in-ember-objects': require('./rules/avoid-leaking-state-in-ember-objects'),
     'avoid-using-needs-in-controllers': require('./rules/avoid-using-needs-in-controllers'),
+    'balanced-braces': require('./rules/balanced-braces'),
```
Can we actually rename this rule to `no-invalid-dependent-keys`? Why?

1. This rule could catch any invalid syntax. Balanced braces is just one part, but this rule can also look for other invalid characters, invalid position of `@each`, `[]`, etc. Note that you don't have to implement everything now but we should name the rule so that it can be expanded afterward.
2. This name would follow the pattern of the existing rule name [no-duplicate-dependent-keys](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/no-duplicate-dependent-keys.md).

TheMBTH commented:
Sorry about the multiple commits & pushes. This is the second time I push and PR on Github :confused: 
I also missed something obvious about recommended rules because I created the tests before running generator-eslint.

I struggled a bit to have the CI pass so I will another issue to improve documentation for "newbies" like me :stuck_out_tongue:

bmish commented on tests/lib/rules/balanced-braces.js:
```diff
@@ -0,0 +1,105 @@
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 18,
+          nodeType: 'CallExpression',
```
This should be `type` not `nodeType`.

bmish commented on tests/lib/rules/balanced-braces.js:
```diff
@@ -0,0 +1,105 @@
+    '{ test: computed.or("foo.bar.name", "foo.bar.place") }',
+    '{ test: computed.and("foo.bar.name", "foo.bar.place") }',
+    '{ test: Ember.computed.filterBy("a", "b", false) }',
+  ],
```
We need some valid examples that allowed unbalanced braces in function calls that are not Ember's `computed` function to make sure those are still allowed.

TheMBTH replied:
I was not sure about what you meant exactly. I simply added function calls containing String arguments that have unbalanced braces.
Did I understand correctly ?

bmish replied:
For example, this should be a valid test case, to ensure we avoid false positives:
```
someOtherFunctionCall('a.{{something}')
```

TheMBTH replied:
Ok great, this is what I did
`    'myFunction("{ key: string }, saving,test}", "b}", false)',
    'myFunction("A string containing curly braces {}}")',`
Should I add your test case or change something in mine? 
Thx :smile_cat:

bmish replied:
Yours is fine, it looks like it achieves the goal of testing for false positives.

TheMBTH replied:
:+1:  I resolve this conversation then. I think I applied all the changes you recommended.

bmish commented:
@TheMBTH can you remove the unrelated commits?

TheMBTH commented:
I rebased and removed the unrelated commits. I rarely rebase so I am sure about the result.
I wanted to have README updated so I pulled master. How should we keep applying changes from master while we are still working on a PR?
thx :pray:

bmish commented on docs/rules/no-invalid-dependent-keys.md:
```diff
@@ -0,0 +1,42 @@
+
+## Rule Details
+
+This rule aims to avoid invalid dependent keys in computed properies.
```
typo: `properties`

bmish commented on lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,61 @@
+            context.report({
+              node,
+              message: ERROR_MESSAGE,
+              loc: {
```
`loc` should be removed. Instead, the `node` that is reported should be string literal node with the violation.

bmish commented on docs/rules/no-invalid-dependent-keys.md:
```diff
@@ -0,0 +1,42 @@
+    // Code
+  })
+});
+
```
```suggestion
```

bmish commented on docs/rules/no-invalid-dependent-keys.md:
```diff
@@ -0,0 +1,42 @@
+    // Code
+  })
+});
+
```
```suggestion
```

bmish commented on lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,61 @@
+              },
+            });
+          });
+        }
```
```suggestion
        const argMap = node.arguments.filter(
          arg => types.isLiteral(arg) && typeof arg.value === 'string'
        );

        const problemNodes = argMap.filter(node => {
          const stringValue = node.value;
          const foundBraces = stringValue.match(/[{}]/g) || [];
          const openBraces = foundBraces.filter(c => c === '{').length;
          const closeBraces = foundBraces.filter(c => c === '}').length;
          return openBraces !== closeBraces;
        });

        problemNodes.forEach(problemNode => {
          context.report({
            node: problemNode,
            message: ERROR_MESSAGE,
          });
        });
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 18,
+          type: 'CallExpression',
```
```suggestion
          type: 'Literal',
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 18,
+          type: 'CallExpression',
```
```suggestion
          type: 'Literal',
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 18,
+          type: 'CallExpression',
```
```suggestion
          type: 'Literal',
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 19,
+          type: 'CallExpression',
```
```suggestion
          type: 'Literal',
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 19,
+          type: 'CallExpression',
```
```suggestion
          type: 'Literal',
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          column: 18,
+          type: 'CallExpression',
+          endLine: 1,
+          endColumn: 54,
```
```suggestion
          endColumn: 35,
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+          severity: 1,
+          message: ERROR_MESSAGE,
+          line: 1,
+          column: 18,
```
```suggestion
          column: 37,
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+eslintTester.run('no-invalid-dependent-keys', rule, {
+  valid: [
+    '{ test: computed("a", "b", function() {}) }',
+    '{ test: computed(function() {}) }',
```
```suggestion
    '{ test: computed(function() {}) }',
    '{ test: computed() }',
    '{ test: computed(SOME_VAR, function() {}) }',
    '{ test: computed(...SOME_ARRAY, function() {}) }',
```

bmish commented on tests/lib/rules/no-invalid-dependent-keys.js:
```diff
@@ -0,0 +1,107 @@
+// Tests
+// ------------------------------------------------------------------------------
+
+const eslintTester = new RuleTester();
```
```suggestion
const eslintTester = new RuleTester({ parserOptions: { ecmaVersion: 6, sourceType: 'module' } });
```

TheMBTH commented:
Hi @bmish  .... Sorry for the delay. I didn't know there was a release coming and couldn't find the time. 
Thanks again for the help. I'll make sure to check your updates.

TheMBTH commented:
A few questions regarding your changes:
1 - My tests passed with 'nodeType' because and error needs a type. So the type was actually not checked. It's the message that has a nodeType. Am I correct?
`if (error.type) {
                            assert.strictEqual(message.nodeType, error.type, `Error type should be ${error.type}, found ${message.nodeType}`);
                        }`
1 - The tests still pass without these options. What was the pupose of adding them?
`  parserOptions: {
    ecmaVersion: 6,
    sourceType: 'module',
    ecmaFeatures: { legacyDecorators: true },
  },`

Have a great week-end :smile_cat: 

Cheers

bmish commented:
Thanks for contributing @TheMBTH!

1. Each error object in invalid test cases needs to specify the `type` of node that is reported as the violation. I changed your rule to report violations on specific string literal nodes and then updated the corresponding tests.
2. If you look at the latest master, I added a test to ensure your rule works with the `@computed` decorator, which needs additional `parserOptions`.

TheMBTH commented:
Okay thx.
1- understood
2- I saw the test about `@computed` but when I removed the `parserOptions` to see check how it works, the tests still passed by passing ` parser: require.resolve('babel-eslint')`  only.
I'll look at it on my end to understand.

bmish commented:
I see, if any of the new `parserOptions` are unnecessary, feel free to remove them!
